### PR TITLE
Cluster attraction groups

### PR DESCRIPTION
**Pull request by sfkhalid**
#### Description
In order to place blocks in user-specified floorplan regions, the packing stage of vpr should pack primitives together when they are part of the same floorplan region. This is especially important if the specified floorplan regions are small, since it will be harder to make sure all the necessary primitives are placed within the region if they are packed into many different clustered blocks.

To achieve this goal during packing, attraction groups were introduced to the clustering process. The attraction groups are made up of atoms that are part of the same partition. During clustering, the molecule gain is increased for molecules that have the same attraction group as the cluster (if the cluster has an associated attraction group). This helps ensure that primitives in the same attraction group get packed together.

#### Related Issue
This pull request is part of the work related to issue #932.

### CODE REVIEW

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2065]: Remove or change comment

File: vpr/src/base/vpr_context.h
Comment (sfkhalid) [inline vpr/src/base/vpr_context.h:421]: Might be better in the clustering context.

Comment (sfkhalid) [reply vpr/src/base/vpr_context.h:421]: Clear the attraction groups after clustering and comment that you are doing so. (Would lead to wasted memory if no one else needs it after clustering).

Comment (sfkhalid) [reply vpr/src/base/vpr_context.h:421]: Actually would be better if you create it early (during prepacking) and then pass it around by reference to the routines that need it during clustering. Can find some different structures that are all passed around during clustering and group them together so that you don't have to add it individually to too many parameter lists. Can call it during cl_optimize or something along those lines.

Can create it in a header file (probably cluster.h). Can just make it a struct with no functions (or maybe have a constructor and destructor if they are all created and destroyed and the same time).

File: vpr/src/pack/attraction_groups.cpp
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.cpp:3]: for a short function like this that just returns an int, could make it an inline function in the header file

File: vpr/src/pack/attraction_groups.cpp
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.cpp:7]: return by reference, not by value

File: vpr/src/pack/attraction_groups.cpp
Comment (sfkhalid) [reply vpr/src/pack/attraction_groups.cpp:3]: make all the short ones in this file inline in the header file

File: vpr/src/pack/attraction_groups.cpp
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.cpp:11]: see if you should also add the atom in the other data structure (the vector of atoms that are part of an attraction group)

Comment (sfkhalid) [reply vpr/src/pack/attraction_groups.cpp:11]: may not need these accessors, can also just give the callers direct access to the data structures

File: vpr/src/pack/attraction_groups.cpp
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.cpp:35]: make this a default constructor, can get num atoms from that atom context

File: vpr/src/pack/attraction_groups.h
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.h:14]: Add high-level comment. What is the purpose of attraction group, comment each member of the attraction group and say what it is used for. For example, why do I need a bool that tells me whether the region is size one

File: vpr/src/pack/attraction_groups.h
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.h:32]: add short comment to all of these, ex. setters and getters below

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:305]: could put atom_molecules, cluster_placement_stats_ptr, and the attraction groups data structure in a struct together and pass that around

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:306]: the feasible block array size could go into t_pb_stats with the actual array

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2058]: comment what this is doing

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2069]: use the attraction group gain to update

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:2069]: right now you are updating the blocks that are part of the attraction group multiple times - decide if that is a problem. If it isn't a problem, comment that that is what you are doing. If it is a problem, write code to make sure that you only do this once (i.e. did the cluster attraction group just change) - if you can do this it is good because it can be faster

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2137]: will have these passed in as a parameter as a part of the struct I am creating

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:2137]: can remove the above to-do comment from Vaughn. Or change the comment to a better overall comment for what the function does.

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2164]: The StrongId types (like AttractGroupId) will have AttractGroupId::INVALID(). Use that and get rid of NO_ATTRACTION_GROUP

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2456]: delete this commented out code

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2586]: can change j to indicate that it is an atom within the molecule (i_atom)

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2582]: might be adding the molecule multiple times (because different atoms might be part of the same molecule) -> if not a problem, comment it, if it is -> can have an unordered set to check if it has been added.

Also put this in a subroutine and refactor all of the functions that go through this atom_molecules loop to call the subroutine

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2653]: delete this routine

File: vpr/src/pack/prepack.cpp
Comment (sfkhalid) [inline vpr/src/pack/prepack.cpp:1689]: put this inside the attraction group class. make a default constructor that sets it to invalid/empty. and then put this code in an initialization routine in the class.

Comment (sfkhalid) [reply vpr/src/pack/prepack.cpp:1689]: if you have all the info you need though, just put this routine into the default constructor.

Comment (sfkhalid) [reply vpr/src/pack/prepack.cpp:1689]: if you can return stuff by reference in these routines, do so

File: vpr/src/pack/prepack.h
Comment (sfkhalid) [inline vpr/src/pack/prepack.h:25]: won't need this declaration anymore

**Comment by vaughnbetz** [comment]
Added WIP to the title as this isn't ready for merge yet.
Email from @luujason indicates our understanding of the pb_stats and gains seems correct -- only the top-level (root/cluster) pb_stats is used for gains.  That means @sfkhalid should refactor the code to move the gains-related data out of the pb_stats data structure and into a new cl_optimize data structure (along with the other data we talked about today).  
That could be done in a separate pr if it is easier to get the current code going with the redundant pb_stats loading code for attraction gains -- I suspect that is best as it will minimize the size of each PR.

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:2582]: Does not get added multiple times because add_molecule_to_pb_stats checks whether the molecule is already in feasible blocks

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:309]: feasible_block_array_size is the packer option that specifies the max size of the array. It is different from the number of feasible blocks currently in the array, which is already part of t_pb_stats

**Comment by sfkhalid** [comment]
QoR comparison of this branch vs. master:
[att_groups_qor_comparison.xlsx](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6862570/att_groups_qor_comparison.xlsx)
No significant changes noted in run time.

**Comment by vaughnbetz** [comment]
This branch was created on Sunday, July 11 (or very close to that date).  That's well before @tangxifan 's recent RRGraphBuilder changes were merged (that was four days ago, or July 18).  Both the sanitized push test and the sanitized pull request test on vtr_reg_basic are failing due to out of memory.  Is the push test running only on Sarah's branch (without pulling the changes into master) while the pull request test is running on the master with Sarah's changes pulled in?  If so, the fact that both are failing suggests @tangxifan 's recent changes are not the culprit.  (I'm not quite sure of how the push vs. pull request tests differ, or if in fact they're just duplicates of each other).

**Comment by vaughnbetz** [comment]
S:basic on the push had 7 failures and S:basic on the pull had 6 failures.  So if they're running on the same code, that indicates some contention between jobs that's leading to out of memory errors as it isn't completely predictable.

**Comment by vaughnbetz** [comment]
Odin basic tests on the pull request failed 6 benchmarks and Odin basic tests on the push failed 7 so again some difference.

**Comment by vaughnbetz** [comment]
The pattern of failures in the main body of the Odin II basic regtests looks the same in the push and pull request versions (same tests fail, just sometimes in a slightly different order, presumably due to parallel jobs).

**Comment by vaughnbetz** [comment]
Odin II basic regtests are configured to use the sanitizers as well.  So that looks like a big clue: the two sets of runs we have that use the sanitizers (Odin II & vtr basic) on github are failing with out of memory.  
The strong sanitized runs are not failing, but they run on kokoro which has bigger machines.  I think we have a configuration issue and need to increase the memory on the github runners for the sanitized basic tests, or move them to kokoro.
@tangxifan : adding your name to make sure you get this :).

File: vpr/src/pack/attraction_groups.h
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.h:24]: rephrase more precisely - "atoms in the same att group will be highly desirable to be packed together. Will be seen as having higher gain if in the same att group. Right now the att groups are based on floorplan constraints, atoms in the same partition are in the same att group. Right now used for floorplanning, but could be used for on other concepts in the future

File: vpr/src/pack/attraction_groups.h
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.h:31]: change this comment to att group

File: vpr/src/pack/attraction_groups.h
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.h:50]: potentially rename - must_be_packed_in_one_cluster

Comment (sfkhalid) [reply vpr/src/pack/attraction_groups.h:50]: can even just comment this out for now, add it back when something uses it

File: vpr/src/pack/attraction_groups.h
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.h:62]: return by const reference

File: vpr/src/pack/attraction_groups.h
Comment (sfkhalid) [inline vpr/src/pack/attraction_groups.h:77]: Atoms with no attraction group will have AttractGroupId::INVALID()

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2057]: write a better overall comment for this routine

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:2057]: routine is called when new molecule added to cluster, updates gain for blocks in marked blocks array

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2129]: Write better overall comment - a molecule has been added to the cluster, this updates stats on that cluster such as used pins, etc. and makes calls to make sure gains of other atoms are adjusted to reflect the new cluster content

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2165]: To-Do comment - allow more than one attraction group

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:2165]: otherwise you are overwriting the attraction group that was previously in the cluster

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:2165]: can have a vector of attraction groups for each cluster

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:2605]: add Jason Luu's name so people would know who to ask about this

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:305]: make the functions you are adding static (add to declaration and definition) as well - makes them file scope

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [inline vpr/src/pack/cluster.cpp:311]: these should also technically be static

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:305]: also add comments in the definitions (to match the other routines)

Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:305]: add comments for ones you have figured out as well

**Comment by sfkhalid** [comment]
As a note, I ran the sanitized basic tests locally, and they are failing locally as well. I ran the test on both this branch (cluster_attraction_groups) and the master branch, and there are failing tests on both branch. The tests are failing with the exit code 23.

**Comment by tangxifan** [comment]
> As a note, I ran the sanitized basic tests locally, and they are failing locally as well. I ran the test on both this branch (cluster_attraction_groups) and the master branch, and there are failing tests on both branch. The tests are failing with the exit code 23.

@sfkhalid Can you share the log files on the locally failed tests? I can look into it and figure it how to patch. Thanks!

**Comment by sfkhalid** [comment]
> > As a note, I ran the sanitized basic tests locally, and they are failing locally as well. I ran the test on both this branch (cluster_attraction_groups) and the master branch, and there are failing tests on both branch. The tests are failing with the exit code 23.
> 
> @sfkhalid Can you share the log files on the locally failed tests? I can look into it and figure it how to patch. Thanks!

Sure, below are the log files of some of the failing tests as well as summary of which tests failed.

[master_ch_intrinsics_vpr.txt](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6889109/master_ch_intrinsics_vpr.txt)
[master_diffeq1_vpr.txt](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6889110/master_diffeq1_vpr.txt)
[master_mkPktMerge_vpr.txt](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6889114/master_mkPktMerge_vpr.txt)
[master_single_ff_vpr.txt](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6889115/master_single_ff_vpr.txt)
[master_single_wire_vpr.txt](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6889116/master_single_wire_vpr.txt)
[master_run_summary.txt](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6889117/master_run_summary.txt)

These are all from the run done on the master branch. I've compared with the run that was done on the attraction groups branch and so far all the failures seem to be the same.

**Comment by tangxifan** [comment]
@sfkhalid Just checked the reports. It seems to be a TBB issue. I double checked the log file of [S:Basic test](https://pipelines.actions.githubusercontent.com/k4WhX565mV5YoOwLfWUFDkcVVIQ2u8tSLZAJkQP8ExBUWpSugE/_apis/pipelines/1/runs/1827/signedlogcontent/22?urlExpires=2021-07-28T22%3A23%3A22.3594732Z&urlSigningMethod=HMACV1&urlSignature=RhB1mfJxq3v1pkOqcXetbOEyPvSVW%2BDXNdErqG7O5Yc%3D). The TBB is actually disabled in these tests. In my local tests, TBB was off as well. 

If you can turn off TBB, we may bypass this issue and see if we see the same issue as the Github runner shows.
Or we may consider to add some tests for TBB

**Comment by sfkhalid** [comment]
@tangxifan Did your local sanitized basic tests pass with TBB off?

**Comment by tangxifan** [comment]
> @tangxifan Did your local sanitized basic tests pass with TBB off?

@sfkhalid Yes. That is what I see on my side.

**Comment by sfkhalid** [comment]
> > @tangxifan Did your local sanitized basic tests pass with TBB off?
> 
> @sfkhalid Yes. That is what I see on my side.

Ok, I see, thank you. I'll try running without tbb on my side and see what happens

**Comment by vaughnbetz** [comment]
There is also a suppression file for the sanitizers that the regtests use.  Are you using that @sfkhalid ?  I believe there are a few issues in the Intel TBB library (small memory leaks, some object accesses that the sanitizer complains about) that we can't fix, so we suppress those errors.

**Comment by sfkhalid** [comment]
> There is also a suppression file for the sanitizers that the regtests use. Are you using that @sfkhalid ? I believe there are a few issues in the Intel TBB library (small memory leaks, some object accesses that the sanitizer complains about) that we can't fix, so we suppress those errors.

@vaughnbetz No, I wasn't aware of that file. Where is it located/how do I enable it for the reg test?

**Comment by vaughnbetz** [comment]
The leak sanitizer suppression file is at https://github.com/verilog-to-routing/vtr-verilog-to-routing/tree/master/vpr/lsan.supp

But looking at it, it doesn't have TBB suppression, just one suppression for graphics.  There is also a valgrind suppression file, and it also doesn't have TBB suppresion (but perhaps doesn't need one). @kmurray : should we make an lsan.supp file with the TBB memory leaks and invalid object messages suppressed?  If you already know those are spurious warnings it seems like we should make one.

**Comment by sfkhalid** [comment]
A summary of some test cases:
[Attraction Groups Testing.docx](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6901327/Attraction.Groups.Testing.docx)
These test cases were done to see whether using attraction groups during packing was giving any added benefit to helping VPR succeed with floorplan constraints.

File: vpr/src/pack/cluster.cpp
Comment (vaughnbetz) [reply vpr/src/pack/cluster.cpp:2609]: This is marked as resolved but I don't see Jason Luu's name added?

File: vpr/src/pack/cluster.cpp
Comment (vaughnbetz) [inline vpr/src/pack/cluster.cpp]: No comment on what this function does.  Should have one.
I think this is a file scope function so should be static (I don't think it is in the header).
Should apply the same logic to all added functions: comment their use, and make them static if their declaration is not in the .h file. Comment their use in the .h file if they are declared there, or in the .cpp file (at the definition) if declared static in the .cpp file.
Should also declare all static functions near the top of the .cpp file (where the other static functions are declared).

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
Seems that not all the comments from our in-zoom review have been addressed, although the comments are marked resolved.  Please address before merging.  
Should update qor spreadsheet too so it checks the final code (after all updates).

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp]: Sure, I'll add a comment to this function and the other add_cluster_molecule functions.
I made all of the add_cluster_molecule_candidates_by_x functions static and added the declaration for this new one at the top of the file, I'm not sure why that's not showing here. I can see it on this pull request if I go on under files changed though.

File: vpr/src/pack/cluster.cpp
Comment (sfkhalid) [reply vpr/src/pack/cluster.cpp:2609]: It is added at the end of the comment, I'll move it to the beginning though so it's more clear

File: vpr/src/pack/cluster.cpp
Comment (vaughnbetz) [reply vpr/src/pack/cluster.cpp]: Thanks.  Maybe I clicked on the wrong link and wasn't looking at the latest commit.

**Comment by sfkhalid** [comment]
QoR spreadsheet for code comparison after all changes:
[att_groups_qor_comparison_2.xlsx](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6909198/att_groups_qor_comparison_2.xlsx)
No significant changes were noted across the categories between the master branch and the attraction groups branch.

**Comment by vaughnbetz** [comment]
Looks good; merging.

**Comment by sfkhalid** [comment]
> Looks good; merging once CI goes green unless you tell me there is a reason to wait @sfkhalid

Nope, should be good to merge once the tests pass.

**Comment by sfkhalid** [comment]
@vaughnbetz The tests passed, I think this is good to merge now

**Pull request merged**
